calico_cni_k8s_test.go

```diff
@@ -842,95 +847,229 @@ var _ = Describe("CalicoCni", func() {
+			}{
+				{
+					description: "old-style inline subnet",
+					config: `
+					{
+					  "cniVersion": "%s",
+					  "name": "net6",
+					  "nodename_file_optional": true,
+					  "type": "calico",
+					  "etcd_endpoints": "http://%s:2379",
+					  "datastore_type": "%s",
+					  "ipam": {
+					    "type": "host-local",
+					    "subnet": "usePodCidr"
+					  },
+					  "kubernetes": {
+					   "k8s_api_root": "http://127.0.0.1:8080"
+					  },
+					  "policy": {"type": "k8s"},
+					  "log_level":"debug"
+					}`,
+					expectedV4Routes: []string{
+						regexp.QuoteMeta("default via 169.254.1.1 dev eth0"),
+						regexp.QuoteMeta("169.254.1.1 dev eth0 scope link"),
+					},
+					unexpectedRoute: regexp.QuoteMeta("10."),
+				},
+				{
+					// This scenario tests IPv4+IPv6 without specifying any routes.
+					description: "new-style with IPv4 and IPv6 ranges, no routes",
+					config: `
+					{
+					  "cniVersion": "%s",
+					  "name": "net6",
+					  "nodename_file_optional": true,
+					  "type": "calico",
+					  "etcd_endpoints": "http://%s:2379",
+					  "datastore_type": "%s",
+					  "ipam": {
+					    "type": "host-local",
+					    "ranges": [
+					       [
+					          {
+					            "subnet": "usePodCidr"
+					          }
+					       ],
+					       [
+					          {
+					            "subnet": "dead:beef::/96"
+					          }
+					       ]
+					    ]
+					  },
+					  "kubernetes": {
+					   "k8s_api_root": "http://127.0.0.1:8080"
+					  },
+					  "policy": {"type": "k8s"},
+					  "log_level":"debug"
+					}`,
+					expectedV4Routes: []string{
+						regexp.QuoteMeta("default via 169.254.1.1 dev eth0"),
+						regexp.QuoteMeta("169.254.1.1 dev eth0 scope link"),
+					},
+					expectedV6Routes: []string{
+						"dead:beef::. dev eth0  metric 256",
+						"fe80::/64 dev eth0  metric 256",
+						"default via fe80::ecee:eeff:feee:eeee dev eth0  metric 1024",
+						"ff00::/8 dev eth0  metric 256",
+					},
+					unexpectedRoute: regexp.QuoteMeta("10."),
+				},
+				{
+					// In this scenario, we use a lot more of the host-local IPAM plugin.  Namely:
+					// - we use multiple ranges, one of which is IPv6, the other uses the podCIDR
+					// - we add custom routes, which override our default 0/0 and ::/0 routes.
+					description: "new-style with IPv4 and IPv6 ranges and routes",
+					config: `
+					{
+					  "cniVersion": "%s",
+					  "name": "net6",
+					  "nodename_file_optional": true,
 					  "type": "calico",
 					  "etcd_endpoints": "http://%s:2379",
 					  "datastore_type": "%s",
 					  "ipam": {
 					    "type": "host-local",
-						"subnet": "usePodCidr"
+					    "ranges": [
+					       [
+					          {
+					            "subnet": "usePodCidr"
+					          }
+					       ],
+					       [
+					          {
+					            "subnet": "dead:beef::/96"
+					          }
+					       ]
+					    ],
+					    "routes": [
+					      {"dst": "10.123.0.0/16"},
```

Comment:
Can there be a gateway field here?  I understand that we don't use it, but should it be here to check that we can ignore it?